File: 프로그래머스/lv1/64061. 크레인 인형뽑기 게임/크레인 인형뽑기 게임.java
import java.util.*;

class Solution {
    public int solution(int[][] board, int[] moves) {
        // Stack 객체 생성
        Stack<Integer> stack = new Stack<>();
        
        int answer = 0;
        
        for(int i = 0; i < moves.length; i++) {
            for(int j = 0; j < board.length; j++) {
                int doll = board[j][moves[i]-1];
                // board에 인형이 존재하는 경우
                if(doll != 0) {
                    // Stack이 비어있지 않은 경우
                    if(!stack.isEmpty()) {
                        if(stack.peek() == doll) {
                            board[j][moves[i]-1] = 0;
                            stack.pop();
                            answer = answer + 2;
                            break;
                        } else {
                            stack.push(doll);
                            board[j][moves[i]-1] = 0;
                            break;
                        }
                    } else {
                        stack.push(doll);
                        board[j][moves[i]-1] = 0;
                        break;
                    }
                }
            }
        }
        
        return answer;
    }
}

File: 프로그래머스/1/64061. 크레인 인형뽑기 게임/크레인 인형뽑기 게임.js
const solution = (board, moves) => {
    let bucket = []; // 뽑은 인형을 담을 배열
    let count = 0; // 터뜨려진 인형의 갯수

    // 이동 배열 순회
    for(let i = 0; i < moves.length; i++) {
        let doll = 0; // 뽑은 인형
        // board행 순회
        for(let j = 0; j < board.length; j++) {
            if(board[j][moves[i] - 1] === 0) {
                continue;
            } else {
                doll = board[j][moves[i] - 1];
                board[j][moves[i] - 1] = 0;
                break;
            }
        }

        if(doll === 0) continue;

        // 바구니 제일 위에 있는 인형이 뽑은 인형과 같다면
        if(bucket.at(-1) === doll) {
            bucket.pop();
            count += 2;
        } else {
            bucket.push(doll);
        }
    }

    return count;
}